File: julia/nucleotide-count/nucleotide-count.jl
"""
    count_nucleotides(strand)

The frequency of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
  dna = Dict('A'=>0,'C'=>0,'G'=>0,'T'=>0)

  for i in strand
    try
      dna[i] += 1
    catch
      throw(DomainError("Invalid Key"))
    end
  end

  return dna
end

File: ocaml/nucleotide-count/nucleotide_count.ml
(*
Given a string representing a DNA sequence, count how many of each nucleotide
is present. If the string contains characters that aren't A, C, G, or T then it
is invalid and you should signal an error.

For example:

"GATTACA" -> 'A': 3, 'C': 1, 'G': 1, 'T': 2
"INVALID" -> error
*)

open Base

let empty = Map.empty (module Char)

let count_nucleotide s c = Ok (String.to_list s |> List.count ~f:(fun f -> phys_equal f c))

let count_nucleotides s =
  List.map ~f:(fun f -> count_nucleotide s f) ['A'; 'C'; 'G'; 'T']